src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -13,12 +14,125 @@ namespace Aspire.Hosting.ApplicationModel;
+            token = timeoutCts.Token;
+        }
+        try
+        {
+            var targetStates = !string.IsNullOrEmpty(targetState) ? new[] { targetState } : [KnownResourceStates.Running];
+            return WaitForResourceAsync(resourceName, targetStates, token);
+        }
+        finally
+        {
+            timeoutCts?.Dispose();
+        }
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach one of the specified states. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches one of the specified target states. If the resource
+    /// is already in the target state, the method will return immediately.<br/>
+    /// If the resource doesn't reach one of the target states within 60 seconds, it will throw <see cref="TaskCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resource.</param>
+    /// <param name="targetStates">The set of states to wait for the resource to transition to one of. See <see cref="KnownResourceStates"/> for common states.</param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    public Task<string> WaitForResourceAsync(string resourceName, IEnumerable<string> targetStates)
+    {
+        var timeoutCts = new CancellationTokenSource(s_defaultWaitForResourceStateTimeout);
+        try
+        {
+            return WaitForResourceAsync(resourceName, targetStates, timeoutCts.Token);
+        }
+        finally
+        {
+            timeoutCts.Dispose();
+        }
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach one of the specified states. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches one of the specified target states. If the resource
+    /// is already in the target state, the method will return immediately.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resource.</param>
+    /// <param name="targetStates">The set of states to wait for the resource to transition to one of. See <see cref="KnownResourceStates"/> for common states.</param>
+    /// <param name="cancellationToken">A cancellation token that cancels the wait operation when signalled.</param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    public async Task<string> WaitForResourceAsync(string resourceName, IEnumerable<string> targetStates, CancellationToken cancellationToken)
+    {
+        var watchCts = CancellationTokenSource.CreateLinkedTokenSource(_applicationStopping, cancellationToken);
+        var watchToken = watchCts.Token;
+        try
+        {
+            await foreach (var resourceEvent in WatchAsync(watchToken).WithCancellation(watchToken).ConfigureAwait(false))
```

Comment:
Why do we pass the token twice

Reply:
You're right, that's not required. It was related to the other comment about how WatchAsync doesn't combine with the app stopping token by default but this is doing it anyway.